test/tests/functional/pbs_node_rampdown.py

```diff
@@ -6362,3 +6415,774 @@ def test_release_nodes_shared_server_restart_immed(self):
+        # self.match_accounting_log('E', subjob1, self.job1_exec_host_esc,
+        #                          self.job1_exec_vnode_esc, "6gb",
+        #                          8, 3, self.job1_place, self.job1_sel_esc)
+
+    def test_release_job_states(self):
+        """
+        Release nodes on jobs in various states; Q, H, S, W
+        """
+
+        # Submit a regular job that cannot run
+        job = Job(TEST_USER)
+        a = {'Resource_List.ncpus': 100}
+        j = Job(TEST_USER, a)
+        jid = self.server.submit(j)
+
+        self.server.expect(JOB, {'job_state': 'Q'}, id=jid)
+
+        # Release nodes from a queued job
+        cmd = [self.pbs_release_nodes_cmd, '-j', jid, self.n4]
+        ret = self.server.du.run_cmd(self.server.hostname, cmd,
+                                     sudo=True)
+        self.assertNotEqual(ret['rc'], 0)
+
+        self.server.delete(jid, wait=True)
+
+        # Submit a held job and try releasing the node
+        j1 = Job(TEST_USER, {ATTR_h: None})
+        jid1 = self.server.submit(j1)
+        self.server.expect(JOB, {'job_state': 'H'}, id=jid1)
+
+        cmd = [self.pbs_release_nodes_cmd, '-j', jid1, self.n4]
+        ret = self.server.du.run_cmd(self.server.hostname, cmd,
+                                     sudo=True)
+        self.assertNotEqual(ret['rc'], 0)
+        self.server.delete(jid1, wait=True)
+
+        # Submit a job in W state and try releasing the node
+        mydate = int(time.time()) + 120
+        mytime = self.convert_time('%m%d%H%M', str(mydate))
+        j2 = Job(TEST_USER, {ATTR_a: mytime})
+        jid2 = self.server.submit(j2)
+        self.server.expect(JOB, {'job_state': 'W'}, id=jid2)
+
+        cmd = [self.pbs_release_nodes_cmd, '-j', jid2, self.n4]
+        ret = self.server.du.run_cmd(self.server.hostname, cmd,
+                                     sudo=True)
+        self.assertNotEqual(ret['rc'], 0)
+        self.server.delete(jid2, wait=True)
+
+    def test_release_finishjob(self):
+        """
+        Test that releasing vnodes on finished jobs will fail
+        also verify the updated schedselect on a finished job
+        """
+
+        self.server.manager(MGR_CMD_SET, SERVER,
+                            {'job_history_enable': "true"}, sudo=True)
+
+        jid = self.create_and_submit_job('job1_5')
+        self.server.expect(JOB, {'job_state': "R"}, id=jid)
+
+        # Release hostC
+        cmd = [self.pbs_release_nodes_cmd, '-j', jid, self.n7]
+        ret = self.server.du.run_cmd(self.server.hostname, cmd)
+        self.assertEqual(ret['rc'], 0)
+
+        # Submit another job and make sure it is
+        # picked up by hostC
+        j = Job(TEST_USER,
+                {'Resource_List.select': "1:host=" + self.hostC})
+        jid2 = self.server.submit(j)
+        self.server.expect(JOB, {'job_state': "R",
+                                 "exec_host": "centos3/1"}, id=jid2)
```

Comment:
I'm assuming that centos3 is a hostname?  Putting static host names in PTL tests is bad since the test will fail on any other host.  Maybe use self.momX.short_hostname?

Reply:
Oops! thanks for catching that. It was left by mistake.